firedrake/projection.py

```diff
@@ -35,72 +81,129 @@ def project(v, V, bcs=None, mesh=None,
     If ``V`` is a :class:`.Function` then ``v`` is projected into
     ``V`` and ``V`` is returned. If `V` is a :class:`.FunctionSpace`
     then ``v`` is projected into a new :class:`.Function` and that
-    :class:`.Function` is returned.
+    :class:`.Function` is returned."""
 
-    The ``mesh`` and ``form_compiler_parameters`` are currently ignored."""
-    from firedrake import function
+    val = Projector(v, V, bcs=bcs, solver_parameters=solver_parameters,
+                    form_compiler_parameters=form_compiler_parameters).project()
+    val.rename(name)
+    return val
 
-    if isinstance(V, functionspaceimpl.WithGeometry):
-        ret = function.Function(V, name=name)
-    elif isinstance(V, function.Function):
-        ret = V
-        V = V.function_space()
-    else:
-        raise RuntimeError(
-            'Can only project into functions and function spaces, not %r'
-            % type(V))
 
-    if isinstance(v, expression.Expression):
-        shape = v.value_shape()
-        # Build a function space that supports PointEvaluation so that
-        # we can interpolate into it.
-        if isinstance(V.ufl_element().degree(), tuple):
-            deg = max(V.ufl_element().degree())
-        else:
-            deg = V.ufl_element().degree()
+class Assigner(object):
+    def __init__(self, source, target):
+        self.source = source
+        self.target = target
 
-        if v.rank() == 0:
-            fs = functionspace.FunctionSpace(V.mesh(), 'DG', deg+1)
-        elif v.rank() == 1:
-            fs = functionspace.VectorFunctionSpace(V.mesh(), 'DG',
-                                                   deg+1,
-                                                   dim=shape[0])
+    def project(self):
+        self.target.assign(self.source)
+        return self.target
```

Comment:
Used in the case where the source and target are both functions in the same space.